src/theory/quantifiers/sygus/sygus_reconstruct.h

```diff
@@ -0,0 +1,320 @@
+#include "theory/quantifiers/sygus/term_database_sygus.h"
+
+namespace CVC4 {
+namespace theory {
+namespace quantifiers {
+
+/** SygusReconstruct
+ *
+ * This class implements an algorithm for reconstructing a given term such that
+ * the reconstructed term is equivalent to the original term and its syntax
+ * matches a specified grammar.
+ *
+ * Goal: find a term g in sygus type T that is equivalent to builtin term t.
+ *
+ * rcons(t, T) : Sol<k_0>
+ * {
+ *   // a list of triples to reconstruct into Sygus type T, where
+ *   // a labeled term is of the form (k, t, s), where k is a skolem of type T,
+ *   // t is a builtin term of the type encoded by T, and s is a possibly null
+ *   // sygus term of type T representing the solution.
+ *   Obs<T>
+ *
+ *   // We define Term<k> and Sol<k> for each skolem k such that Obs<T> contains
+ *   // entries (k, Term<k>, Sol<k>)
+ *
+ *   CandSols<k> : a list of possible solutions, for each label k. Whenever
+ *                 there is a successful match, it's added to this list.
+ *   Pool<T> : a list of enumerated terms for each sygus type
+ *
+ *   let k_0 be a fresh skolem of sygus type T
+ *
+ *   Obs<T> += {(k_0, t, null)}
+ *
+ *   Pool<T> = {}
+ *
+ *   while Sol<k_0> == null
+ *     // enumeration phase
+ *     for each T
+ *       s[z] = nextEnum(T_{z})
+ *       if (s[z] is ground)
+ *         if exists (k', t', s') in Obs<T> s.t. |=_X t' = toBuiltIn(s[z])
+ *           markSolved(k', s[z])
+ *         else
+ *           let k be a new variable of type : T
+ *           Obs<T> += (k, toBuiltIn(s[z]), s[z])
+ *       else if no s' in Pool<T> s.t.
+ *             rewrite(toBuiltIn(s')) * sigma = rewrite(toBuiltIn(s))
+ *           Pool<T> += s[z]
+ *           for each (k, t, null) in Obs<T>
+ *             Obs' += matchNewObs(k, s[z])
+ *     // match phase
+ *     while Obs' != {}
+ *       Obs'' = {}
+ *       for each (k, t, null) in Obs' // s = null for all triples in Obs'
+ *         Obs<T> += (k, t, null)
+ *         for each s[z] in Pool<T>
+ *           Obs'' += matchNewObs(k, s[z])
+ *       Obs' = Obs''
+ * }
+ *
+ * matchNewObs(k, s[z]) : Obs'
+ * {
+ *   u = rewrite(toBuiltIn(s[z]))
+ *   if match(u, t) == {toBuiltin(z) -> t'}
+ *     if forall t' exists (k'', t'', s'') in Obs<T> s.t. |=_X t'' = t'
+ *       markSolved(k, s{z -> s''})
+ *     else
+ *       let k' be a new variable of type : typeOf(z)
+ *       CandSol<k> += s{z -> k'}
+ *       Obs'<typeOf(z)> += (k', t', null)
+ * }
+ *
+ * markSolved(k, s)
+ * {
+ *   if Sol<k> != null
+ *     return
+ *   if s is not ground
+ *     instantiate s with arbitrary sygus datatype values
+ *   Sol<k> = s
+ *   CandSols<k> += s
+ *   S = {k}      // a stack to process
+ *   while S != {}
+ *     k' = pop(S)
+ *     for all k'' in CandSols keys
+ *       for all s'[k'] in CandSols<k''>
+ *         s'{k' -> Sol<k'>}
+ *         if s' contains free variables not in Obs
+ *           instantiate s with arbitrary sygus datatype values
```

Comment:
Do you mean s'?  Also this can be merged with the above line e.g. "instantiate all free variables in s' with arbitrary values".

Reply:
Done. To some degree...